### DISCUSSION TITLE
[python-nrepl] Loading the right site-packages

### DISCUSSION THREAD

**Discussion by nicolas-graves** [nrepl/nrepl#362] [General]
Hi nrepl!

Thanks a lot! I was continuing my python-nrepl, and I have some difficulty to tackle one subject, I need some feedback. 

The issue is about how to properly pass the information about the python site-packages to import between the client and the server. 

1) The server can be started from a place where it has no idea which packages will be necessary to run the client code.
2) Using a generic client such as `emacs-arei` make language-specific operations / environment injections difficult.
3) I don't see a dedicated ops to request a `PYTHONPATH` env variable from the server to the client.

How is this kind of thing handled in the Clojure version of the package? 

I have a solution which consists in reading from the direnv `.envrc` after finding the `.envrc` (since the client passes the filename to the server, locally the server can deduce the location of .envrc and execute it), but that doesn't seem to be a fitting solution when the client and the server are not on the same machine.

Thanks again!

**Comment by alexander-yakushev** [comment]
I don't quite understand what issue you are trying to solve, to be honest. If this is Python-specific, I can't really help you much. In Clojure implementation, the server and the client are separate Java/Clojure processes with their own dependencies, so there is no problem like the one you describe?.. I'm not sure if I'm even thinking in the correct direction, sorry.

**Comment by nicolas-graves** [comment]
It's indeed written as Python-specific, I'll try to ask more general questions and deduce what I need from there. 

Let's say you have client code which relies on a dependency that is not installed on the server. How is the code executed on the server in that case?

**Comment by alexander-yakushev** [comment]
It doesn't. nREPL briefly experimented with the idea of side-loading client-supplied dependencies to the server, but there was not enough buy-in from the clients (editors), so the feature was recently scrapped.It is assumed that the server starts with all the necessary dependencies on the classpath. I still like the idea of being able to add dependencies in the runtime, but the strategy how do it needs more thought.

**Comment by nicolas-graves** [comment]
Thanks for your clear answer. I don't think a python nrepl server without dependencies is of much use in my use-case, so I'm going to stick with my local-only solution (It won't work running an isolated or remote server). 

A last question that might be able to help me: how would you make the server ask the client for a env variable or a configuration item? I know there's the stdin operation (although never used it on my server), but do you know of a use-case where the client answers automatically for a specific stdin request? 

[edit] I'm also interested about the details or the side-loading client-supplied dependencies you describe, it there a design document or something I can read about that? Thanks!

**Comment by alexander-yakushev** [comment]
I don't remember if the sideloading design was fully documented before deleted. You may search the repository for "sideload" to pick up some breadcrumbs in the PRs.

> how would you make the server ask the client for a env variable or a configuration item?

CIDER clients mostly pass all the relevant configuration to the server in each request, or once to set it statefully. Can you describe the flow you envision where you would need the server to ask the client for configuration?

**Comment by nicolas-graves** [comment]
To put it simply, I want to side-load client-supplied dependencies. 
I have a way to do this IFF the server has access to the host filesystem (~ running locally), because the server can deduce the location of project-specific files for the `filename`. This code is a big dirty hack. What I would find relevant is for the server to ask this info (in my case the `PYTHONPATH` env variable, or could be entire files to load) to the client, and then the client answers with required info. 

But I think with a light config handling (not implemented yet in `emacs-arei` IIRC) sent with every request to the server, this could indeed suit my needs.

**Comment by alexander-yakushev** [comment]
You can implement a middleware op something like `load-path` that the client sends to the server once along with the necessary information, and the server will then do the loading and maybe save the parameters in its session storage or something. If you simplify the setup to only run the server and the client on the same machine, then it becomes pretty trivial to do. The main source of complexity for nREPL's side-loading implementation was to ship the deps themselves over the wire. If you don't need to do that, it becomes much easier.

**Comment by nicolas-graves** [comment]
Thanks @alexander-yakushev  
I will give more info when I have time to tackle this
Maybe you can convert the issue to a Github "discussion" not to let that rot in issues?
